src/System.Reflection.DispatchProxy/src/System.Reflection.DispatchProxy.csproj

```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <StrongNameKeyId>Open</StrongNameKeyId>
+    <IsPackable>false</IsPackable>
+    <VersionPrefix>4.7.1</VersionPrefix>
+    <VersionPrefix Condition="'$(IsPackable)' == 'true'">4.8.0</VersionPrefix>
+    <AssemblyVersion>4.0.6.0</AssemblyVersion><!-- NO-INCREMENT: This version is frozen for .NET Standard and .NETCoreApp because the assembly ships inbox. -->
+    <AssemblyVersion Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">4.0.6.0</AssemblyVersion>
+    <AssemblyVersion Condition="'$(IsPackable)' == 'true' and $([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">4.0.7.0</AssemblyVersion>
+    <PackageValidationBaseLineVersion>4.7.1</PackageValidationBaseLineVersion>
+    <!-- this library depends on Ref.Emit which is not available for netstandard -->
+    <GeneratePlatformNotSupportedAssemblyMessage Condition="'$(TargetFramework)' == 'netstandard2.0'">SR.PlatformNotSupported_ReflectionDispatchProxy</GeneratePlatformNotSupportedAssemblyMessage>
```

Comment:
`GeneratePlatformNotSupportedAssemblyMessage`? That's a dotnet/runtime thing which doesn't exist here.

Reply:
The comment above dates [back to 2.0](https://github.com/dotnet/corefx/blob/d6173e069a9bcedfdfd7f4f41e67d23f67157b61/src/System.Reflection.DispatchProxy/src/System.Reflection.DispatchProxy.csproj#L9) before Ref.Emit had it's own PNSE's in the package:

Reflection.Emit's package was brought back later in https://github.com/dotnet/corefx/commit/1e72f59677547df9b82fa642fd4f443c6768d743 and was made available to NETStandard https://www.nuget.org/packages/System.Reflection.Emit.  It's a PNSE itself.  We could decide to make the netstandard2.0 implementation a real one and just let the Ref.Emit PNSE do the throwing.

It would just change the callstack slightly in the exception slightly, but I think it would still have the same behavior.

That said - we still might want to decide what we do for PNSE's in this repo.  I think we said before that we could just check in the generated PNSE code, does that sound right @ViktorHofer?

Reply:
> That said - we still might want to decide what we do for PNSE's in this repo. I think we said before that we could just check in the generated PNSE code, does that sound right @ViktorHofer?

Yes in general that makes sense. Here I only see three public APIs so it might make sense to put the PNSEs directly into the existing code with ifdefs.

Reply:
There are 4 APIs: The abstract class itself, a protected constructor, an abstract method, and a static method.

Questions:
- How do I ifdef the type and the abstract method?
- Do I use `NETSTANDARD2_0_OR_GREATER` or just `NETSTANDARD2_0` or `NETSTANDARD`?

```cs
public abstract class DispatchProxy // 1
{
    protected DispatchProxy() // 2
    {
#if NETSTANDARD2_0_OR_GREATER
            throw new PlatformNotSupportedException(); // I could do this too, but this is protected
#endif
    }
    protected abstract object Invoke(MethodInfo targetMethod, object[] args); // 3
    public static T Create<T, TProxy>() // 4
        where TProxy : DispatchProxy
    {
#if NETSTANDARD2_0_OR_GREATER
            throw new PlatformNotSupportedException();
#else
            return (T)DispatchProxyGenerator.CreateProxyInstance(typeof(TProxy), typeof(T));
#endif
    }
}
```

Reply:
I _think_ the abstract needs to be completely surrounded, and not be abstract in netstandard, like this?:

```cs
#if NETSTANDARD2_0_OR_GREATER
        protected object Invoke(MethodInfo targetMethod, object[] args)
        {
            throw new PlatformNotSupportedException();
        }
#else
        protected abstract object Invoke(MethodInfo targetMethod, object[] args);
#endif
```

Reply:
And maybe for the type itself there's nothing to do.

Reply:
Oh and I suspect that I need to use:
```cs
throw new PlatformNotSupportedException(SR.PlatformNotSupported_ReflectionDispatchProxy);
```

If we want to keep this property in the csproj:
```csproj
<GeneratePlatformNotSupportedAssemblyMessage Condition="'$(TargetFramework)' == 'netstandard2.0'">
SR.PlatformNotSupported_ReflectionDispatchProxy</GeneratePlatformNotSupportedAssemblyMessage>

```

Reply:
Oh, ignore that last comment. The property is from runtime.

Reply:
I think I did the right thing, so I'll resolve this conversation. Please let me know if I missed anything.